syntax = "proto3";

package envoy.extensions.regex_engine.v3;

import "google/protobuf/wrappers.proto";

import "udpa/annotations/status.proto";

option java_package = "io.envoyproxy.envoy.extensions.regex_engine.v3";
option java_outer_classname = "GoogleRe2Proto";
option java_multiple_files = true;
option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/regex_engine/v3;regex_enginev3";
option (udpa.annotations.file_status).package_version_status = ACTIVE;

// [#protodoc-title: Google RE2]

// Google's `RE2 <https://github.com/google/re2>`_ regex engine. The regex string must adhere to
// the documented `syntax <https://github.com/google/re2/wiki/Syntax>`_. The engine is designed
// to complete execution in linear time as well as limit the amount of memory used.
//
// Envoy emits two stats for tracking the program size of regexes: the histogram `re2.program_size`,
// which records the program size, and the counter `re2.exceeded_warn_level`, which is incremented
// each time the program size exceeds the warn level threshold.
message GoogleRE2 {
  // This field controls the RE2 "program size" which is a rough estimate of how complex a
  // compiled regex is to evaluate. A regex that has a program size greater than the configured
  // value will fail to compile. In this case, the configured max program size can be increased
  // or the regex can be simplified. If not specified, the default is 100.
  google.protobuf.UInt32Value max_program_size = 1;

  // A regex that has a program size greater than the configured value will generate a warning log.
  // If not specified, no warning log will be logged.
  google.protobuf.UInt32Value warn_program_size = 2;
}
